# Raízes de Equações Transcendentais


- Encontrar a raiz da função, significa: encontrar os valores para que f(x) seja igual a zero.
- Gráficamente é encontrar o ponto em que a função cruza o eixo x
![imagem](https://i.ibb.co/CvfN5d0/image.png)

- A figura mostra o comportamento da função f(p) = 2^(1-p)-10^(-5)
    - f(17) = 0.53*10^(-5)
    - f(18) = -0.24*10^(-5)
    - A raiz está estre 17 e 18
    
- Existem métodos que auxiliam em achar o zero da função:

## Método da Bissecção

- Seja f(x) uma função contínua no intervalo [a,b]. Se f(a)f(b) < 0, então f(x) tem pelo menos uma raiz em (a,b)

__Passo a passo:__
- Encontrar um intervalo [a,b]
- Secciona o intervalo no seu ponto médio x = (a+b)/2
- Se x for uma solução aceitável para o valor da raiz, pare
- Senão, volte ao inicio para verificar se a raiz está em [a,x] ou em [x,b]. Redefina o intervalo [a,b] e volte ao passo 1
- O critério de parada é o erro relativo:
    - O algoritmo deve parar quando estimações sucessivas estão próximas o suficiente entre si. 
    - Seja X_(k+1) a estimativa do valor da raiz na iteração (k+1) e X_k a estimativa na iteração anterior (k)
    - Er = abs (X_(k+1) - X_k / X(k+1)
    - O algoritmo deve parar uma vez que x_(k+1) coincida em pelo menos p algarismos com x_k
    - Er = abs (X_(k+1) - X_k / X(k+1) <= 10^(-p)
    

In [1]:
def Bis(a,b,p,N): #Define a função bissecação
    erro=1 #Erro começa com 1
    cont=1 #Contador começando em 1
    
    #Enquando erro for maior que a precisão E
    #cont for menor que o número maximo de iterações
    
    while erro > 10**(-p) and cont<N: 
        rn=(b+a)/2 #Raiz nova será (b+a)/2
        #Se f(rn)*f(a) for menor que 0
        #Existe raíz entre rn e a
        if func(rn)*func(a)<0: 
            b=rn # b recebe o valor de rn
            erro=abs(b-a)/abs(b) #calculo do erro
        #Se não, existe raíz entre rn e b    
        else:
            a=rn #a recebe o valor de rn
            erro=abs(a-b)/abs(a) #calculo do erro
        
        cont=cont+1 #incrementa o contador 
    
    print(cont, rn) #mostra quantas iterações foram feitas
    # e o valor mais próximo da raiz que mais aproxima a função de zero

##### Exemplo:  
Ache a raiz da função X^3-9X+3

In [3]:
def func(x): 
    return x**3-9*x+3 #Definindo a função do exemplo

In [4]:
#Chamando a função, a=0, b=1, p=10^-3, e N=50
Bis(0,1,3,50) 


13 0.337646484375


Após 13 iterações, o valor que deixa a função do exemplo mais perto de ser zero é 0.337646484375

## Método da Newton

- Conhecendo-se um valor x0 próximo à raiz, pode-se aproximar a função por uma reta tangente à f(x) em x0.

__Passo a passo:__
- Analisar a função para determinar um valor de x0 próximo a raiz
- Encontrar a derivada de f(x)
- Calcular estimativas consecutivas para a raiz usando a expressão: x_(k+1) = x_k - f(x_k)/f'(x_k)
- Repetir o passo 3 até que se tenha obtido a precisão desejada

    

In [5]:
def Newton(x0,p,N):
    erro=1
    ra=x0
    cont=1
    rn=ra
    while erro > 10**(-p) and cont<N:
        ra=rn
        rn=ra-func(ra)/dfunc(ra)
        cont=cont+1
        erro=abs(rn-ra)/abs(rn)
    
    print(cont, rn)

#### Exemplo

Ache a raiz da função X^3-2X^2-3X+10

In [6]:
def func(x):
    return x**3-2*x**2-3*x+10 #Função da equação do exemplo

def dfunc(x):
    return 3*x**2-4*x-3 #Função com a derivada da equação do exemplo



In [7]:
Newton(1,3,50)

19 -2.0000000001846723


Após 19 iterações, a o valor que mais aproxima a função do exemplo a zero é -2.0000000001846723

## Método da Secante

- Aproximar a derivada no ponto (x1, f(x1)) pelo coeficiente angular da reta secante que passa pelos pontos (x1,f(x1)) e (x0, f(x0)).

__Passo a passo:__
- Analisar a função para determinar dois valores, x0 e x1, próximos à raiz
- Calcular estimativas consecutivas para a raiz usando a expressão: x_(k+1) = x_k-f(x_k)* (x_k-x_(k-1))/(f(x_k)-f(x_(k-1)))
- Repetir o passo 2 até que se tenha obtido a precisão desejada

In [8]:
def Secante(x0,x1,p,N):
    erro=1
    ra1=x0
    cont=1
    rn=x1
    while erro > 10**(-p) and cont<N:
        ra2=ra1
        ra1=rn
        rn=ra1-func(ra1)*(ra2-ra1)/(func(ra2)-func(ra1))
        cont=cont+1
        erro=abs(rn-ra1)/abs(rn)
    
    print(cont, rn)

#### Exemplo

Ache a raiz da função X^3-2X^2-3X+10

In [9]:
Secante(1,3,3,50)

3 -2.0


Após 3 iterações, o valor mais próximo de zerar a função foi -2.0